### PR TITLE
Add tests for #37309 and #58472

### LINKED ISSUE #37309: GetTypeInfo() on a type in a foreach loses top level nullability
Given:

    foreach (string? s in Foo())

If I pass the `string?` to GetTypeInfo we would expect to get a string that's either annotated as nullable or it's flow state is MaybeNull. (Debatable which it should be…). But we don't get either of those and it seems it's just dropped.

### LINKED ISSUE #56498: Unexpected formatting warning for a space between a `new` modifier and a tuple type syntax.
Microsoft Visual Studio Enterprise 2022 Int Preview (64-bit)
Version 17.0.0 Preview 5.0 [31713.378.main]

image

The warning goes away if the space is removed:
image

Expected: No warning when the space is present, a warning when it is missing.

Note: this issue is referenced in source.

### LINKED ISSUE #58472: Array index from end in ValueTuple assignment throws IndexOutOfRangeException
### Description

Assignment from tuple destructuring into a FromEnd index of a generic array throws IndexOutOfRangeException. It looks like `stelem` is called with the Index's raw value, instead of a computed value from the array length.

### Reproduction Steps

```
M(new int[1]);

void M<T>(T[] a)
{
    (a[0], a[^1]) = (default, default);
}
```

### Expected behavior

Assignment succeeds

### Actual behavior

IndexOutOfRangeException

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

.NET 6.0.101

### Other information

_No response_

### PR DESCRIPTION
**Pull request by jcouv**
Closes #37309 
Closes #58472

### CODE REVIEW

File: eng/Versions.props [hunk 1]
```diff
@@ -31,7 +31,7 @@
     <MicrosoftCodeAnalysisNetAnalyzersVersion>6.0.0-rc1.21366.2</MicrosoftCodeAnalysisNetAnalyzersVersion>
     <MicrosoftCodeAnalysisTestingVersion>1.1.1-beta1.21480.3</MicrosoftCodeAnalysisTestingVersion>
     <!-- CodeStyleAnalyzerVersion should we updated together with version of dotnet-format in dotnet-tools.json -->
-    <CodeStyleAnalyzerVersion>4.0.0-3.final</CodeStyleAnalyzerVersion>
+    <CodeStyleAnalyzerVersion>4.0.0-6.final</CodeStyleAnalyzerVersion>
```
Comment (jcouv) [inline eng/Versions.props:34]: @JoeRobich Could you help with identifying the corresponding version for `dotnet-format`? Thanks

**Review by 333fred — APPROVED** [review]

**Comment by AlekseyTs** [comment]
It looks like the title of the PR is misleading. It claims that the PR is a test only change, but in fact:
- it changes product code
- it makes some infrastructure changes

I suggest splitting PRs into 3. At least the infrastructure changes should be separated, I think. Also all PRs should have titles that accurately reflect the nature of changes.

**Comment by jcouv** [comment]
I don't think the PR contains any meaningful product code or infrastructure change. I'm opportunistically trying to close issues that are low-hanging fruits. Making more PRs seems like just overhead given the tiny size of the change, but that's fine:

https://github.com/dotnet/roslyn/pull/58810
https://github.com/dotnet/roslyn/pull/58808

**Pull request merged**